### DISCUSSION TITLE
Exhaustive matches on sequences

### DISCUSSION THREAD

**Discussion by sirrus233** [microsoft/pyright#5252] [Q&A] [answered]
I noticed that the following code reports a non-exhaustive pattern match, and I could not understand why:

```
my_list = [i for i in range(10)]

match my_list:
    case []:
        ...
    case [first]:
        ...
    case [first, *rest]:
        ...
```

Is this a bug, something beyond the intended capabilities of pyright, or is there a value I haven't thought of that would fail to match any of these cases?

**Comment by erictraut** [comment, marked as answer]
Exhaustive matching is performed in type checkers using a technique called [type narrowing](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing). In your code sample, there's no opportunity for type narrowing based on the supplied patterns. The type of the subject expression `my_list` starts as `list[int]`. After the first case statement, it's still a `list[int]`. There's no type in the Python type system that means "a `list[int]` whose length is not zero". Likewise, the type cannot be narrowed after the second or third case statements, so the final type is still `list[int]`. It would need to be `never` for the match statement to be proven exhaustive. The `list` type in the Python system doesn't have any notion of length.

So no, this isn't a bug in pyright. This is a limitation of the Python type system. Mypy and other type checkers will likewise be unable to prove that this case statement is exhaustive.
